### ISSUE TITLE
Proposal: Switch to PSR-7 / PSR-15 for HTTP requests to the Magento application

### ISSUE THREAD

**Issue by marcusirgens** [magento/community-features#292]
TL;DR:I propose switching from Magento's request/response objects to PSR-7 objects and PSR-15 handlers.

While Magento 2 already has it's own bespoke interfaces for representing requests and responses, the PHP community has since the inception of Magento 2 created two solid standards for messaging and server interaction with requests. These are:

- [The HTTP Message Interface](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-7-http-message.md)
- [HTTP Server Request Handlers](https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-15-request-handlers.md)

While simple, these interfaces provide a solid and universal framework for handling and manipulating HTTP requests.

There are multiple reasons for this suggestion:
1. Magento will at some point need to change its request/response models, as the current implementation prevents the application from processing multiple requests without bootstrapping over.
2. There's already a rich and expanding ecosystem providing PSR-7 and PSR-15 components such as middleware, handlers, routers, decorators, etc
3. Magento 2's Router component needs a rewrite, and I believe it's important to at least have a discussion about this before effort is spent on rewriting a router system for the bespoke req/res objects.

Interestingly, I think this has the potential to reduce the complexity for multiple things that are somewhat difficult to handle in Magento 2 at the moment:

1. Authentication (ie. in the Admin backend) could be made more explicit using middleware
2. "Union" interfaces can be used to support handlers (eg. magento interfaces extending the PSR-15 for routers), while removing a lot of junk and complex data structures

As for implementation, I believe this should be possible to implement for 2.5 with some work done to provide information to the community, and that the current implementation can be deprecated by 2.6 or 2.7.

**Comment by bfontaine** [comment]
Has there been anything new on this since 2020?

**Comment by marcusirgens** [comment]
> Has there been anything new on this since 2020?

@bfontaine I no longer work with Magento or the PHP ecosystem, and have not done so since 2020. If this is still relevant, I suggest that you or other active contributors open a new issue.
